prow/plugins/plugins.go

```diff
@@ -107,54 +110,154 @@ type PluginAgent struct {
 	PluginClient
 
 	mut sync.Mutex
-	// Repo (eg "k/k") -> list of handler names.
-	ps map[string][]string
+	// pc maps repo (eg "k/k") to enabled
+	// plugins and their config for that repo.
+	pc map[string]Config
 }
 
+// Config holds the plugin configuration
+// for a repository or an organization
+type Config struct {
+	EnabledPlugins []string `json:"-"`
+
+	Trigger *Trigger `json:"trigger,omitempty"`
+	Heart   *Heart   `json:"heart,omitempty"`
+	Label   *Label   `json:"label,omitempty"`
+}
+
+type Trigger struct {
+	// TrustedOrg is the org whose members' PRs will be automatically built
+	// for PRs to the above repos.
+	TrustedOrg string `json:"trusted_org,omitempty"`
+}
+
+type Heart struct {
+	// Adorees is a list of GitHub logins for members
+	// for whom we will add emojis to comments
+	Adorees []string `json:"adorees,omitempty"`
+}
+
+type Label struct {
+	// SigOrg is the organization that owns the
+	// special interest groups tagged in this repo
+	SigOrg string `json:"sig_org,omitempty"`
+}
+
+// pluginConfig is used as a deserialization
+// target to validate the plugins requested
+// by users in the plugin configuration YAML
+type pluginConfig map[string]interface{}
+
 // Load attempts to load config from the path. It returns an error if either
 // the file can't be read or it contains an unknown plugin.
 func (pa *PluginAgent) Load(path string) error {
-	b, err := ioutil.ReadFile(path)
+	reader, err := os.Open(path)
+	if err != nil {
+		return err
+	}
+	// We want to deserialize to two different targets
```

Comment:
This really sucks but I wanted to keep the same API for plugin lists as we had before... and Go has no simple way to do strict JSON deserialization.